File: app/src/test/java/com/weiqi/engine/RulesTest.kt
package com.weiqi.engine

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class RulesTest {

    private fun newGame(size: Int = 9, komi: Double = 7.5) =
        GameState.newGame(GameConfig(boardSize = size, komi = komi))

    private fun play(state: GameState, r: Int, c: Int): GameState {
        val res = Rules.apply(state, MoveIntent(MoveType.PLACE_STONE, Point(r, c)))
        assertTrue("expected accepted at ($r,$c) but got $res", res is MoveResult.Accepted)
        return (res as MoveResult.Accepted).newState
    }

    private fun pass(state: GameState): GameState {
        val res = Rules.apply(state, MoveIntent(MoveType.PASS))
        return (res as MoveResult.Accepted).newState
    }

    @Test fun basicPlacement() {
        var s = newGame()
        s = play(s, 4, 4)
        assertEquals(CellState.BLACK, s.board.get(Point(4, 4)))
        assertEquals(StoneColor.WHITE, s.currentPlayer)
    }

    @Test fun rejectOccupied() {
        var s = newGame()
        s = play(s, 4, 4)
        val res = Rules.apply(s, MoveIntent(MoveType.PLACE_STONE, Point(4, 4)))
        assertEquals(MoveResult.Rejected(MoveResult.Reason.OCCUPIED), res)
    }

    @Test fun rejectOutOfBounds() {
        val s = newGame(9)
        val res = Rules.apply(s, MoveIntent(MoveType.PLACE_STONE, Point(9, 0)))
        assertEquals(MoveResult.Rejected(MoveResult.Reason.OUT_OF_BOUNDS), res)
    }

    @Test fun singleStoneCapture() {
        // White stone at (1,1) surrounded by black on all four sides.
        var s = newGame()
        s = play(s, 1, 1) // B
        // To make white go on (2,2) trapped, we need a different setup.
        // Easier: black plays around a white stone in the corner.
        s = newGame()
        s = play(s, 0, 1) // B
        s = play(s, 0, 0) // W (corner)
        s = play(s, 1, 0) // B  -> captures W at (0,0)
        assertEquals(CellState.EMPTY, s.board.get(Point(0, 0)))
        assertEquals(1, s.capturesByBlack)
    }

    @Test fun rejectSuicide() {
        // Surround (0,0) with white stones, then black plays (0,0): suicide.
        var s = newGame()
        s = play(s, 5, 5) // B (away)
        s = play(s, 0, 1) // W
        s = play(s, 5, 6) // B
        s = play(s, 1, 0) // W
        // Black to move at (0,0) — surrounded by white, no captures.
        val res = Rules.apply(s, MoveIntent(MoveType.PLACE_STONE, Point(0, 0)))
        assertEquals(MoveResult.Rejected(MoveResult.Reason.SUICIDE), res)
    }

    @Test fun captureBeatsApparentSuicide() {
        // White group with 1 liberty; black plays into that liberty and captures, not suicide.
        var s = newGame()
        // Build white group at (0,0)-(0,1) with last liberty (0,2).
        s = play(s, 1, 0) // B
        s = play(s, 0, 0) // W
        s = play(s, 1, 1) // B
        s = play(s, 0, 1) // W
        s = play(s, 1, 2) // B
        s = play(s, 8, 8) // W away
        // Now black plays (0,2): captures white group.
        s = play(s, 0, 2) // B
        assertEquals(CellState.EMPTY, s.board.get(Point(0, 0)))
        assertEquals(CellState.EMPTY, s.board.get(Point(0, 1)))
        assertTrue(s.capturesByBlack >= 2)
    }

    @Test fun simpleKoBlocksImmediateRecapture() {
        // Classic ko shape on 9x9.
        var s = newGame()
        s = play(s, 2, 2) // B
        s = play(s, 2, 3) // W
        s = play(s, 3, 1) // B
        s = play(s, 3, 4) // W
        s = play(s, 4, 2) // B
        s = play(s, 4, 3) // W
        s = play(s, 3, 3) // B at center; W at (3,2) will be the ko stone
        s = play(s, 3, 2) // W captures B at (3,3)
        // Now black tries to immediately recapture at (3,3) -> ko violation.
        val res = Rules.apply(s, MoveIntent(MoveType.PLACE_STONE, Point(3, 3)))
        assertEquals(MoveResult.Rejected(MoveResult.Reason.KO_VIOLATION), res)
    }

    @Test fun doublePassEntersScoring() {
        var s = newGame()
        s = pass(s)
        s = pass(s)
        assertEquals(GameStatus.SCORING, s.status)
    }

    @Test fun resignEndsGame() {
        var s = newGame()
        val res = Rules.apply(s, MoveIntent(MoveType.RESIGN))
        assertTrue(res is MoveResult.Accepted)
        assertEquals(GameStatus.RESIGNED, (res as MoveResult.Accepted).newState.status)
    }

    @Test fun handicapPlacesBlackStones() {
        val s = GameState.newGame(GameConfig(boardSize = 9, handicap = 2))
        var blacks = 0
        for (r in 0 until 9) for (c in 0 until 9)
            if (s.board.get(Point(r, c)) == CellState.BLACK) blacks++
        assertEquals(2, blacks)
        assertEquals(StoneColor.WHITE, s.currentPlayer)
    }
}

File: test/domain/rules_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:go_game/src/domain/game_state.dart';
import 'package:go_game/src/domain/models.dart';
import 'package:go_game/src/domain/rules.dart';

void main() {
  GameState newGame({int size = 9, double komi = 7.5}) =>
      GameState.newGame(GameConfig(boardSize: size, komi: komi));

  GameState play(GameState state, int r, int c) {
    final res = Rules.apply(state, MoveIntent.place(Point(r, c)));
    expect(res.isAccepted, isTrue, reason: 'expected accepted at ($r,$c)');
    return res.newStateAs<GameState>();
  }

  GameState pass(GameState state) {
    final res = Rules.apply(state, const MoveIntent.pass());
    return res.newStateAs<GameState>();
  }

  test('basic placement', () {
    var s = newGame();
    s = play(s, 4, 4);
    expect(s.board.cellAt(const Point(4, 4)), CellState.black);
    expect(s.currentPlayer, StoneColor.white);
  });

  test('reject occupied', () {
    var s = newGame();
    s = play(s, 4, 4);
    final res = Rules.apply(s, const MoveIntent.place(Point(4, 4)));
    expect(res.isAccepted, isFalse);
    expect(res.reason, MoveRejection.occupied);
  });

  test('reject out of bounds', () {
    final s = newGame(size: 9);
    final res = Rules.apply(s, const MoveIntent.place(Point(9, 0)));
    expect(res.isAccepted, isFalse);
    expect(res.reason, MoveRejection.outOfBounds);
  });

  test('single stone capture', () {
    var s = newGame();
    s = play(s, 0, 1); // B
    s = play(s, 0, 0); // W (corner)
    s = play(s, 1, 0); // B captures W at (0,0)
    expect(s.board.cellAt(const Point(0, 0)), CellState.empty);
    expect(s.capturesByBlack, 1);
  });

  test('reject suicide', () {
    var s = newGame();
    s = play(s, 5, 5); // B (away)
    s = play(s, 0, 1); // W
    s = play(s, 5, 6); // B
    s = play(s, 1, 0); // W
    final res = Rules.apply(s, const MoveIntent.place(Point(0, 0)));
    expect(res.isAccepted, isFalse);
    expect(res.reason, MoveRejection.suicide);
  });

  test('capture beats apparent suicide', () {
    var s = newGame();
    s = play(s, 1, 0); // B
    s = play(s, 0, 0); // W
    s = play(s, 1, 1); // B
    s = play(s, 0, 1); // W
    s = play(s, 1, 2); // B
    s = play(s, 8, 8); // W
    s = play(s, 0, 2); // B captures white group
    expect(s.board.cellAt(const Point(0, 0)), CellState.empty);
    expect(s.board.cellAt(const Point(0, 1)), CellState.empty);
    expect(s.capturesByBlack >= 2, isTrue);
  });

  test('simple ko blocks immediate recapture', () {
    var s = newGame();
    s = play(s, 2, 2); // B
    s = play(s, 2, 3); // W
    s = play(s, 3, 1); // B
    s = play(s, 3, 4); // W
    s = play(s, 4, 2); // B
    s = play(s, 4, 3); // W
    s = play(s, 3, 3); // B
    s = play(s, 3, 2); // W captures B at (3,3)
    final res = Rules.apply(s, const MoveIntent.place(Point(3, 3)));
    expect(res.isAccepted, isFalse);
    expect(res.reason, MoveRejection.koViolation);
  });

  test('double pass enters scoring', () {
    var s = newGame();
    s = pass(s);
    s = pass(s);
    expect(s.status, GameStatus.scoring);
  });

  test('resign ends game', () {
    final s = newGame();
    final res = Rules.apply(s, const MoveIntent.resign());
    expect(res.isAccepted, isTrue);
    expect(res.newStateAs<GameState>().status, GameStatus.resigned);
  });

  test('handicap places black stones', () {
    final s = GameState.newGame(const GameConfig(boardSize: 9, handicap: 2));
    var blacks = 0;
    for (var r = 0; r < 9; r++) {
      for (var c = 0; c < 9; c++) {
        if (s.board.cellAt(Point(r, c)) == CellState.black) blacks++;
      }
    }
    expect(blacks, 2);
    expect(s.currentPlayer, StoneColor.white);
  });

  test('suicide allowed: own stone is removed and credited to opponent', () {
    var s = GameState.newGame(const GameConfig(
      boardSize: 9,
      ruleset: Ruleset.newZealand,
      allowSuicide: true,
      superkoMode: SuperkoMode.situational,
    ));
    // Surround the corner (0,0) with white so a black stone there is suicide.
    s = play(s, 5, 5); // B somewhere safe
    s = play(s, 0, 1); // W
    s = play(s, 5, 6); // B
    s = play(s, 1, 0); // W

    // Black plays into the corner — under NZ rules this is a legal self-capture.
    final res = Rules.apply(s, const MoveIntent.place(Point(0, 0)));
    expect(res.isAccepted, isTrue, reason: 'self-capture should be legal');
    final next = res.newStateAs<GameState>();
    expect(next.board.cellAt(const Point(0, 0)), CellState.empty,
        reason: 'suicided stone should be removed');
    expect(next.capturesByWhite, 1,
        reason: 'self-captured black stone is credited to white');
  });

  test('positional superko forbids identical board with other player to move',
      () {
    // Sketch: under positional, any prior position is forbidden.
    // We approximate by replaying enough to set up a near-cycle and confirm
    // the superko check fires when board would repeat.
    var s = GameState.newGame(const GameConfig(
      boardSize: 9,
      ruleset: Ruleset.chinese,
      superkoMode: SuperkoMode.positional,
    ));
    // Pass twice in a row would normally enter scoring; instead alternate
    // passes with placements that we then capture to attempt a repeat. The
    // simplest reliable check: an initial empty-board state cannot be returned
    // to by any sequence of placements (stones don't vanish without capture).
    // So instead, drive a single capture and verify the post-capture hash is
    // stored.
    s = play(s, 0, 1); // B
    s = play(s, 0, 0); // W
    s = play(s, 1, 0); // B captures W
    expect(s.previousHashes.length, greaterThan(1));
  });

  test('situational superko: same board with same player-to-move forbidden',
      () {
    // Construct manually: insert a fake "prior" hash matching the situational
    // hash for the current position, and verify the next placement is rejected.
    var s = GameState.newGame(const GameConfig(
      boardSize: 9,
      ruleset: Ruleset.aga,
      superkoMode: SuperkoMode.situational,
    ));
    s = play(s, 4, 4); // B at center; now white to move

    // Suppose white plays at (4,5). Compute the situational hash that would
    // result and seed it into previousHashes to simulate a prior occurrence.
    final placed =
        s.board.setCell(const Point(4, 5), CellState.of(StoneColor.white));
    final fakeHash = Rules.positionHash(
        SuperkoMode.situational, placed, StoneColor.black);
    s = s.copyWith(previousHashes: {...s.previousHashes, fakeHash});

    final res = Rules.apply(s, const MoveIntent.place(Point(4, 5)));
    expect(res.isAccepted, isFalse);
    expect(res.reason, MoveRejection.superkoViolation);
  });

  test('basic-ko-only ruleset does not enforce positional superko', () {
    // Under SuperkoMode.none (Japanese), only the koPoint check applies.
    // Seeding previousHashes with the position-after-move hash should NOT
    // cause rejection.
    var s = GameState.newGame(const GameConfig(
      boardSize: 9,
      ruleset: Ruleset.japanese,
      superkoMode: SuperkoMode.none,
    ));
    final placed =
        s.board.setCell(const Point(4, 4), CellState.of(StoneColor.black));
    final fakeHash =
        Rules.positionHash(SuperkoMode.none, placed, StoneColor.white);
    s = s.copyWith(previousHashes: {...s.previousHashes, fakeHash});

    final res = Rules.apply(s, const MoveIntent.place(Point(4, 4)));
    expect(res.isAccepted, isTrue,
        reason: 'basic-ko-only ruleset should ignore positional repetition');
  });
}
